entitlement/src/main/java/org/killbill/billing/entitlement/api/DefaultEntitlement.java

```diff
@@ -336,51 +335,28 @@ public Entitlement cancelEntitlementWithDate(@Nullable final LocalDate entitleme
-
-                final DateTime billingEffectiveCancelDate = dateHelper.fromLocalDateAndReferenceTimeWithMinimum(billingEffectiveDate, getEventsStream().getSubscriptionBase().getStartDate(), updatedPluginContext.getCreatedDate(), contextWithValidAccountRecordId);
-                try {
-                    if (overrideBillingEffectiveDate) {
-                        getSubscriptionBase().cancelWithDate(billingEffectiveCancelDate, callContext);
-                    } else {
-                        getSubscriptionBase().cancel(callContext);
-                    }
-                } catch (final SubscriptionBaseApiException e) {
-                    throw new EntitlementApiException(e);
-                }
-
-                final DateTime entitlementEffectiveCancelDate = dateHelper.fromLocalDateAndReferenceTimeWithMinimum(entitlementEffectiveDate, getEventsStream().getEntitlementEffectiveStartDateTime(), updatedPluginContext.getCreatedDate(), contextWithValidAccountRecordId);
-                final BlockingState newBlockingState = new DefaultBlockingState(getId(), BlockingStateType.SUBSCRIPTION, DefaultEntitlementApi.ENT_STATE_CANCELLED, KILLBILL_SERVICES.ENTITLEMENT_SERVICE.getServiceName(), true, true, false, entitlementEffectiveCancelDate);
-                final Collection<NotificationEvent> notificationEvents = new ArrayList<NotificationEvent>();
-                final Collection<BlockingState> addOnsBlockingStates = computeAddOnBlockingStates(entitlementEffectiveCancelDate, notificationEvents, callContext, contextWithValidAccountRecordId);
-
-                // Record the new state first, then insert the notifications to avoid race conditions
-                setBlockingStates(newBlockingState, addOnsBlockingStates, contextWithValidAccountRecordId);
-                for (final NotificationEvent notificationEvent : notificationEvents) {
-                    recordFutureNotification(entitlementEffectiveCancelDate, notificationEvent, contextWithValidAccountRecordId);
-                }
-
-                return entitlementApi.getEntitlementForId(getId(), callContext);
-            }
-        };
-
-        return pluginExecution.executeWithPlugin(cancelEntitlementWithPlugin, pluginContext);
     }
     
-
     @Override
     public Entitlement cancelEntitlementWithDate(final DateTime entitlementEffectiveDate, final DateTime billingEffectiveDate,
                                                  final Iterable<PluginProperty> properties, final CallContext callContext) throws EntitlementApiException {
         logCancelEntitlement(log, this, entitlementEffectiveDate, billingEffectiveDate, null, null, null); 
+        return cancelEntitlementWithDate(entitlementEffectiveDate, billingEffectiveDate, true, properties, callContext);
+    }
+    
+    
+    
+    private Entitlement cancelEntitlementWithDate(final DateTime entitlementEffectiveDate, final DateTime billingEffectiveDate, final boolean overrideBillingEffectiveDate, 
```

Comment:
I understand what you are saying and I agree. 

This is the issue:

[This](https://github.com/killbill/killbill/blob/8e93f0a3192b0041a25684eb1190c9861f269366/entitlement/src/main/java/org/killbill/billing/entitlement/api/DefaultEntitlement.java#L353) is the original code, it checks the `overrideBillingEffectiveDate` flag and accordingly invokes  `getSubscriptionBase().cancelWithDate` or `getSubscriptionBase().cancel`. Note that it does not check [billingEffectiveCancelDate](https://github.com/killbill/killbill/blob/8e93f0a3192b0041a25684eb1190c9861f269366/entitlement/src/main/java/org/killbill/billing/entitlement/api/DefaultEntitlement.java#L351) because it always has a non-null value irrespective of the value of the `overrideBillingEffectiveDate` flag. This is the reason I had to pass the `overrideBillingEffectiveDate` flag to the internal `cancelEntitlementWithDate` method. 

I've added a workaround now that compares `billingEffectiveDate` to `callContext.getCreatedDate()` but it does not make the code very readable. LMK if you want me to keep this code or revert to the earlier code (that uses the `overrideBillingEffectiveDate`  flag) or if you think of any other better way to do this.

Reply:
Easier to write the [code](https://github.com/killbill/killbill/commit/e898791db7695a048ac4dd0f09dc2580ade30b55) than explain. What do you think - am I missing something?

Reply:
The change looks good @sbrossie, I merged your code into my branch and got a green build. 

There is one thing though that is bothering me.

Earlier, the code was calculating the `billingEffectiveCancelDate` as follows:

```
final DateTime billingEffectiveCancelDate = dateHelper.fromLocalDateAndReferenceTimeWithMinimum(billingEffectiveDate, getEventsStream().getSubscriptionBase().getStartDate(), pluginContext.getCreatedDate(), contextWithValidAccountRecordId);
```

Note that this uses `getEventsStream().getSubscriptionBase().getStartDate()`. 

Now, we are calculating `billingEffectiveCancelDate` as follows:

```
final DateTime billingEffectiveCancelDate = overrideBillingEffectiveDate ? entitlementEffectiveCancelDate : null;
```

The `entitlementEffectiveCancelDate` is calculated as :

```
final DateTime entitlementEffectiveCancelDate = dateHelper.fromLocalDateAndReferenceTimeWithMinimum(entitlementEffectiveDate, getEventsStream().getEntitlementEffectiveStartDateTime(), pluginContext.getCreatedDate(), contextWithValidAccountRecordId);
```

Note that this uses `getEventsStream().getEntitlementEffectiveStartDateTime()`

So, there is a difference in the way `billingEffectiveCancelDate` is calculated. So, wondering if that would cause any issues in the case where `getEventsStream().getSubscriptionBase().getStartDate()` and `getEventsStream().getEntitlementEffectiveStartDateTime()` return different values?

Reply:
I suppose we could have the following scenario where `overrideBillingEffectiveDate=true` and (input) `entitlementEffectiveDate` < `getEventsStream().getEntitlementEffectiveStartDateTime()` < `getEventsStream().getSubscriptionBase().getStartDate()` and so we end up with a `billingEffectiveCancelDate` < `getEventsStream().getSubscriptionBase().getStartDate()`. I think in this case this would fail [here](https://github.com/killbill/killbill/blob/work-for-release-0.23.x/subscription/src/main/java/org/killbill/billing/subscription/api/user/DefaultSubscriptionBaseApiService.java#L782), but I have not confirmed.

So, to your point, I think it would be best to keep the previous behavior, which means that on [line 342](https://github.com/reshmabidikar/killbill/blob/work-for-release-0.23-1375-change-plan/entitlement/src/main/java/org/killbill/billing/entitlement/api/DefaultEntitlement.java#L342), in the case where `overrideBillingEffectiveDate=true`, we should recompute the effective date(time) using `fromLocalDateAndReferenceTimeWithMinimum` and passing the correct 'min=getEventsStream().getSubscriptionBase().getStartDate()'. Can you add this change then?

Reply:
Do you mean doing the following on Line 342:

```
final DateTime billingEffectiveCancelDate = overrideBillingEffectiveDate ? dateHelper.fromLocalDateAndReferenceTimeWithMinimum(billingEffectiveDate, getEventsStream().getSubscriptionBase().getStartDate(), pluginContext.getCreatedDate(), contextWithValidAccountRecordId) : null;
```

Yes, I believe this will result in the same code as before.

I have made this change now, there is an `e2e` [failure ](https://github.com/killbill/killbill/runs/6641695451?check_suite_focus=true) but I was not able to reproduce this on local, the failing test works fine for me. So, the failure could be due to test flakiness. Could you please rerun the build?